### ISSUE TITLE
[iOS] crash when using GDALOpenEx on GeoJSON file

### ISSUE THREAD

**Issue by rascal222** [OSGeo/gdal#3854]
## Expected behavior and actual behavior.

Try to open a GeoJSON file with a FeatureCollection using `GDALOpenEx` method but application crashes.

After investigation, problem seems to come from `strdup` method returning empty string. We've replaced occurences of `strdup` by `CPLStrdup` in `gdal/ogr/ogrsf_frmts/geojson/libjson/json_object.c` file.

## Steps to reproduce the problem.

```c++
GDALOpenEx(filepath, GDAL_OF_VECTOR, nullptr, nullptr, nullptr);
```

content file example:
```json
{
   "type" : "FeatureCollection",
   "features" : [{
      "type": "Feature",
      "geometry": {
         "type": "Point",
         "coordinates": [125.6, 10.1]
      },
      "properties": {
         "name": "Dinagat Islands"
      }
   }]
}
```

## Operating system

iOS 12+
Xcode 12.4+

## GDAL version and provenance

v2.4.4

**Comment by rouault** [comment]
This is really weird. Do you replicate that with the latest GDAL version, 3.3.0, (we don't support 2.4 any more) ? But we haven't changed much that part of the code.  strdup() returning null should only occur in case of out of memory, and in that case, you'd get an assertion in CPLStrdup(). Could you unroll your changes and see if there's a compiler warning when building the json_object.c file ? Alternatively you may install the libjson-c development package independently (homebrew ? or whatever package management you use) and build GDAL against this external libjson-c

**Comment by rascal222** [comment]
Thank you for your replying. Currently we are not able to test against GDAL 3.x since we have a string dependency on PROJ 3 and we are sticked to GDAL 2.4.
However, we checked the compilation logs and read the following compiler warnings:

```
json_object.c:386:36: warning: implicit declaration of function 'strdup' is invalid in C99 [-Wimplicit-function-declaration]
               lh_table_insert(jso->o.c_object, strdup(key), val);
                                                ^
json_object.c:386:36: warning: incompatible integer to pointer conversion passing 'int' to parameter of type 'void *' [-Wint-conversion]
               lh_table_insert(jso->o.c_object, strdup(key), val);
                                                ^~~~~~~~~~~
./linkhash.h:231:54: note: passing argument to parameter 'k' here
extern int lh_table_insert(struct lh_table *t, void *k, const void *v);
                                                    ^
json_object.c:651:25: warning: implicit declaration of function 'strdup' is invalid in C99 [-Wimplicit-function-declaration]
 jso->o.c_string.str = strdup(s);
                       ^
json_object.c:651:23: warning: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
 jso->o.c_string.str = strdup(s);
                     ^ ~~~~~~~~~
```

In fact we do not get a null pointer when we debug the application but instead we have a dangling pointer and the compiler warnings make sense.
Using lijson-c could be an option, but we still haven't look at the code. We will keep you informed on our progress.

**Comment by rouault** [comment]
Digging into history this has been fixed last year in recent GDAL versions per bb87efed9a8be32116f4e9c3f41250479302f560 (only the changes in ogr/ogrsf_frmts/geojson/libjson/GNUmakefile  and ogr/ogrsf_frmts/geojson/libjson/json_tokener.c  are relevant)
